# Structural Simplification in Popular Music

### Project Overview
This project explores whether commercially successful songs tend to be shorter, lyrically simpler, and more repetitive than the broader population of Spotify songs. It combines Billboard chart data, Spotify audio features, and song lyrics to investigate the relationship between chart success and structural simplification in popular music.

// To be expanded

### Main research question
Do commercially successful songs tend to be shorter, simpler, and more repetitive than non-charting songs?

### Hypotheses
- **H0**: There is no difference in average song duration between Billboard-charting songs and non-charting Spotify songs.
- **H1**: Billboard-charting songs are shorter on average than non-charting Spotify songs.
- **H2**: Billboard-charting songs have more repetitive lyrics than non-charting Spotify songs.
- **H3**: Billboard-charting songs have lower lexical diversity than non-charting Spotify songs.
- **H4**: Shorter duration, higher repetition, and lower lexical diversity are associated with chart success.

Datasets:
https://www.kaggle.com/datasets/imuhammad/audio-features-and-lyrics-of-spotify-songs/code/data

https://www.kaggle.com/datasets/dhruvildave/billboard-the-hot-100-songs

https://www.kaggle.com/datasets/maharshipandya/-spotify-tracks-dataset/code/data

# Data Exploration

In [1]:
# imports 
import pandas as pd

In [6]:
spotify = pd.read_csv("spotify_songs.csv")
billboard_charts = pd.read_csv("charts.csv")
music_data = pd.read_csv("dataset.csv")

In [9]:
print("Spotify data shape:", spotify.shape)
print("Billboard charts data shape:", billboard_charts.shape)
print("Music data shape:", music_data.shape)

Spotify data shape: (18454, 25)
Billboard charts data shape: (330087, 7)
Music data shape: (114000, 21)


In [10]:
print("Spotify columns:", spotify.columns)
print("Billboard charts columns:", billboard_charts.columns)
print("Music data columns:", music_data.columns)

Spotify columns: Index(['track_id', 'track_name', 'track_artist', 'lyrics', 'track_popularity',
       'track_album_id', 'track_album_name', 'track_album_release_date',
       'playlist_name', 'playlist_id', 'playlist_genre', 'playlist_subgenre',
       'danceability', 'energy', 'key', 'loudness', 'mode', 'speechiness',
       'acousticness', 'instrumentalness', 'liveness', 'valence', 'tempo',
       'duration_ms', 'language'],
      dtype='str')
Billboard charts columns: Index(['date', 'rank', 'song', 'artist', 'last-week', 'peak-rank',
       'weeks-on-board'],
      dtype='str')
Music data columns: Index(['Unnamed: 0', 'track_id', 'artists', 'album_name', 'track_name',
       'popularity', 'duration_ms', 'explicit', 'danceability', 'energy',
       'key', 'loudness', 'mode', 'speechiness', 'acousticness',
       'instrumentalness', 'liveness', 'valence', 'tempo', 'time_signature',
       'track_genre'],
      dtype='str')


In [14]:
# sample columns from spotify dataset
spotify.sample(5)

,track_id,track_name,track_artist,lyrics,track_popularity,track_album_id,track_album_name,track_album_release_date,playlist_name,playlist_id,...,loudness,mode,speechiness,acousticness,instrumentalness,liveness,valence,tempo,duration_ms,language
10221,4ezxMulRZWUWA875Ckxf3p,Isle of You,Christon Gray,"Hey Miss Amazing, you blow my mind Beautiful L...",29,5SMxQYp21Vf10jlohsJqII,Even With Evil With Me,2011-07-10,Gospel Neo Soul,2wP2UAYC4Nqb6oUXkMJlAz,...,-9.426,0,0.0573,0.77000,0.000000,0.1530,0.464,132.352,238507,en
4664,1yo16b3u0lptm6Cs7lx4AD,Fight The Power,Public Enemy,"NA Yet our best trained, best educated, best e...",56,0aFNb4RDk2hmKKLa0bzXNz,Fear Of A Black Planet,1990-04-10,Minitruckin Playlist,0VVH2Nzj6kBVGK3WIUQMAw,...,-12.970,1,0.2550,0.00484,0.000002,0.5170,0.415,105.974,282640,en
3834,1nInOsHbtotAmEOQhtvnzP,Stronger (What Doesn't Kill You),Kelly Clarkson,You know the bed feels warmer Sleeping here al...,73,0VmE95pr5TSpZWucfyhO5e,Stronger (Deluxe Version),2011-10-24,The Sound of Post-Teen Pop,10FCW9lj0NdeoYI5VVvVtY,...,-4.282,1,0.0475,0.04600,0.000000,0.1120,0.684,116.044,221947,en
1916,0u4kWPHGLCnNhV1OlDxvpD,Rub You The Right Way - J.G. Classic,Johnny Gill,Pump up the bass Hey Pump up the bass Pump Pum...,0,1NfhsIEQ82XXqSFAPX5cXJ,Johnny Gill (Expanded),1990-04-17,New Jack City,6RLvj5OnEuPzI1P02KwnzK,...,-9.850,0,0.0538,0.01460,0.009740,0.0527,0.609,111.599,335000,en
14551,6alCTRgBPoeAFi2N4il9Q4,Did It Again,Lil Tecca,"We love you, Tecca I got a pack, hit 'em and d...",47,65T18oWoikW2MAilg9j8lW,We Love You Tecca,2019-08-30,Sad Trap,7kzKtw5vug3IIxKfLadi8A,...,-9.929,0,0.1910,0.18500,0.000000,0.1140,0.186,148.775,116240,en


In [12]:
# Sample columns from billboard charts dataset
billboard_charts.sample(10)

,date,rank,song,artist,last-week,peak-rank,weeks-on-board
243939,1975-02-08,53,"Shame, Shame, Shame",Shirley And Company,64.0,53,5
44193,2013-05-25,94,High School,Nicki Minaj Featuring Lil Wayne,89.0,68,6
230651,1977-08-27,52,Devil's Gun,C.J. & Company,62.0,36,15
90369,2004-07-17,70,I Want To Live,Josh Gracin,75.0,70,3
111580,2000-06-24,81,I Disappear,Metallica,76.0,76,7
237724,1976-04-17,38,Happy Music,The Blackbyrds,43.0,38,7
294943,1965-05-01,57,Somebody Else Is Taking My Place,Al Martino,53.0,53,6
209593,1981-09-12,94,You Make My Dreams,Daryl Hall John Oates,81.0,5,20
285236,1967-03-11,50,Jimmy Mack,Martha & The Vandellas,68.0,50,3
127576,1997-05-31,77,One Night At A Time,George Strait,68.0,59,9


In [13]:
# Sample columns form hte music data dataset
music_data.sample(10)

,Unnamed: 0,track_id,artists,album_name,track_name,popularity,duration_ms,explicit,danceability,energy,...,loudness,mode,speechiness,acousticness,instrumentalness,liveness,valence,tempo,time_signature,track_genre
74621,74621,7KF2jnp3QIwcl7QJP5i5nW,Los Hermanos,Bloco Do Eu Sozinho,Adeus Você,40,178893,False,0.770,0.324,...,-10.911,0,0.0371,0.5490,0.000042,0.1040,0.508,104.523,4,mpb
78285,78285,4x6JPITCakOVB58LPv9GRA,Troglauer;Stereoact,Tag am Meer (Stereoact Remix),Tag am Meer - Stereoact Remix,33,234375,False,0.728,0.866,...,-6.023,1,0.1740,0.0217,0.000000,0.2740,0.781,128.019,4,party
87174,87174,1jdAzKbRRrIQ3fpodftKfj,Legião Urbana,Mais Do Mesmo,Geração Coca-Cola,48,141400,False,0.574,0.890,...,-5.789,0,0.0410,0.0820,0.000000,0.0741,0.830,97.900,4,r-n-b
13091,13091,7e2kKEOXRsHJzb2X1uGJwQ,Frankie Knuckles,Beyond The Mix,Party At My House,17,305333,False,0.775,0.737,...,-16.082,0,0.0636,0.0147,0.023600,0.0967,0.855,118.733,4,chicago-house
19769,19769,6pba9PrSRk9hfhtav9blTG,Morgan Evans,Country Car Hits,Country Outta My Girl,0,202613,False,0.448,0.932,...,-4.424,1,0.0976,0.0174,0.000000,0.3470,0.709,187.813,4,country
68596,68596,5ng8nlMTWoDyLL5LhiXDxp,Nacho;Los Mendoza,Para entrar en calor,Happy Happy,0,185360,False,0.797,0.795,...,-5.208,0,0.1620,0.4990,0.000000,0.1640,0.897,96.051,4,latino
35389,35389,44pklUNBAlVDCe4eWsrc5e,Zezo,Zezo (O Príncipe dos Teclados),O Que Eu Faço Amanhã?,43,182600,False,0.750,0.820,...,-2.627,0,0.0288,0.3480,0.000000,0.1120,0.579,111.070,4,forro
107382,107382,4sscDOZCkbLSlDqcCgUJnX,DEVO,Freedom of Choice,Whip It,65,161133,False,0.780,0.869,...,-7.255,1,0.0517,0.0744,0.004000,0.0621,0.957,157.902,4,synth-pop
103961,103961,0gPQTLaqHDgdupKEok7J2x,Craig David,Born to Do It,7 Days,71,235133,False,0.659,0.812,...,-7.499,0,0.0487,0.2300,0.000000,0.0951,0.888,83.014,4,soul
60785,60785,5U7gQ28eWDHl3q1a8pLVay,Elephant Man,Good 2 Go,Head Gone / Genie Dance,20,230640,False,0.637,0.732,...,-6.869,1,0.3580,0.5600,0.000000,0.0933,0.963,115.209,4,j-dance


In [ ]:
# Check for missing values in spotify dataset
spotify.isnull().sum().sort_values(ascending=False)

language                    260
lyrics                      260
energy                        0
duration_ms                   0
tempo                         0
valence                       0
liveness                      0
instrumentalness              0
acousticness                  0
speechiness                   0
mode                          0
loudness                      0
key                           0
track_id                      0
track_name                    0
playlist_subgenre             0
playlist_genre                0
playlist_id                   0
playlist_name                 0
track_album_release_date      0
track_album_name              0
track_album_id                0
track_popularity              0
track_artist                  0
danceability                  0
dtype: int64

In [22]:
# Check for missing values in billboard charts dataset
billboard_charts.isnull().sum().sort_values(ascending=False)

last-week         32312
date                  0
rank                  0
song                  0
artist                0
peak-rank             0
weeks-on-board        0
dtype: int64

In [23]:
# Check for missing values in music data dataset
music_data.isnull().sum().sort_values(ascending=False)

artists             1
album_name          1
track_name          1
Unnamed: 0          0
mode                0
time_signature      0
tempo               0
valence             0
liveness            0
instrumentalness    0
acousticness        0
speechiness         0
key                 0
loudness            0
track_id            0
energy              0
danceability        0
explicit            0
duration_ms         0
popularity          0
track_genre         0
dtype: int64

In [28]:
#checking for duplicates in the different datasets
spotify.duplicated().sum()

np.int64(0)

In [29]:
billboard_charts.duplicated().sum()

np.int64(0)

In [30]:
music_data.duplicated().sum()

np.int64(0)

In [31]:
spotify['track_id'].duplicated().sum()

np.int64(0)

In [32]:
music_data['track_id'].duplicated().sum()

np.int64(24259)

In [34]:
# checking for duplicate song-artist pairs in Spotify dataset
spotify[['track_name', 'track_artist']].duplicated().sum()


np.int64(1862)

In [35]:
# checking for duplicate song-artist pairs in Billboard
billboard_charts[['song', 'artist']].duplicated().sum()

np.int64(300406)

In [44]:
((spotify['lyrics'] == 'NA') | (spotify['lyrics'] == 'N/A') | (spotify['lyrics'] == 'NaN')).sum()

np.int64(0)

In [46]:
(spotify['lyrics'] == 'NaN').sum()

np.int64(0)

In [42]:
spotify['lyrics'].sample(10)

2027     NA Si el ritmo te lleva a mover la cabeza Ya e...
51       Fuckers in school telling me, always in the ba...
11516    I've seen life through the eyes of a different...
8539     Because you move me Baby won't you prove me Wr...
63       NA Uuh-uh Uuh-uh Uuh-uh Uuh-uh Yeah, mi chiedo...
104      Feelin' the song, party done, shadows walkin' ...
13466    NA Could it be you calling me down, down, down...
8867                                                   NaN
17208    I'm lying alone with my head on the phone Thin...
11819    I wanted you to know, I love the way you laugh...
Name: lyrics, dtype: str

In [47]:
spotify['lyrics'].isna().sum()


np.int64(260)

In [49]:
spotify['language'].value_counts(dropna=False)

language
en     15405
es      1705
NaN      260
de       241
pt       176
it       133
pl        76
nl        72
fr        65
tl        63
da        40
id        34
ko        29
vi        24
ro        16
sw        16
so        16
cy        12
no        10
hi        10
ja         8
et         7
tr         5
sv         5
af         5
ru         4
sq         3
hr         3
ar         2
cs         2
ca         2
el         2
hu         1
fi         1
sk         1
Name: count, dtype: int64

In [48]:
spotify[spotify[['track_name', 'track_artist']].duplicated(keep=False)] \
    .sort_values(['track_name', 'track_artist']) \
    [['track_id', 'track_name', 'track_artist', 'playlist_name', 'track_popularity', 'duration_ms']] \
    .head(30)

,track_id,track_name,track_artist,playlist_name,track_popularity,duration_ms
13914,5xGxl3Iux0worDz9tLbaDn,#1 Stunna,Big Tymers,Dirty South Rap Classics by DJ HOTSAUCE,0,281960
17699,7kBdenpTmgWfAz9clmBcok,#1 Stunna,Big Tymers,Hip-Hop Drive,48,281960
14354,65NTcXUtOb27NHKQ4fAcw0,'39 - 2011 Mix,Queen,I didn’t know perm stood for permanent (wave),60,210800
14560,6aNP9GlBi3VHPXl7w3Qjr9,'39 - 2011 Mix,Queen,The Queen - La Discografia Completa,46,210800
11623,4xkOaSrkexMciUUogZKVTS,'Till I Collapse,Eminem,Gangster Rap | 100 % Gangster,83,297787
16450,6yr8GiTHWvFfi4o6Q5ebdT,'Till I Collapse,Eminem,Gangster Rap,16,297893
5413,2dLWb2s1QtgpzAGOaBVa0h,1 Question,E-40,Neo-Soul,51,250081
11397,4UipF3IyxsyStTstiwGHhV,1 Question,E-40,Chixtape 5 - Tory Lanez,42,250081
9499,41SwdQIX8Hy2u6fuEDgvWr,10%,KAYTRANADA,Contemporary Urban,70,186680
16030,6Ti0coAmnSsv3Wx09RhoY6,10%,KAYTRANADA,New R&B‏‏​​ ‍,64,186684


In [51]:
spotify_en = spotify[(spotify['language'] == 'en') & (spotify['lyrics'].notna())].copy()
spotify_en.shape

(15405, 25)

In [52]:
spotify_en[['track_name', 'track_artist']].duplicated().sum()

np.int64(1679)

In [54]:
spotify_en_unique = spotify_en.drop_duplicates(subset=['track_name', 'track_artist']).copy()
spotify_en_unique.shape

(13726, 25)

In [55]:
spotify_en_unique[['track_name', 'track_artist']].duplicated().sum()

np.int64(0)

In [58]:
billboard_song_level = billboard_charts[['song', 'artist']].drop_duplicates().copy()
billboard_song_level.shape

(29681, 2)

In [59]:
spotify_en_unique['track_name_norm'] = spotify_en_unique['track_name'].str.lower()
spotify_en_unique['track_artist_norm'] = spotify_en_unique['track_artist'].str.lower()

In [60]:
spotify_en_unique['track_name_norm'] = spotify_en_unique['track_name_norm'].str.replace(r"\(.*?\)", "", regex=True)
spotify_en_unique['track_name_norm'] = spotify_en_unique['track_name_norm'].str.replace(r"\[.*?\]", "", regex=True)
spotify_en_unique['track_name_norm'] = spotify_en_unique['track_name_norm'].str.replace(r"feat\.?.*", "", regex=True)
spotify_en_unique['track_name_norm'] = spotify_en_unique['track_name_norm'].str.replace(r"featuring.*", "", regex=True)
spotify_en_unique['track_name_norm'] = spotify_en_unique['track_name_norm'].str.replace(r"ft\.?.*", "", regex=True)
spotify_en_unique['track_name_norm'] = spotify_en_unique['track_name_norm'].str.replace(r"[^a-z0-9\s]", "", regex=True)
spotify_en_unique['track_name_norm'] = spotify_en_unique['track_name_norm'].str.replace(r"\s+", " ", regex=True).str.strip()

In [61]:
spotify_en_unique['track_artist_norm'] = spotify_en_unique['track_artist_norm'].str.replace(r"\(.*?\)", "", regex=True)
spotify_en_unique['track_artist_norm'] = spotify_en_unique['track_artist_norm'].str.replace(r"\[.*?\]", "", regex=True)
spotify_en_unique['track_artist_norm'] = spotify_en_unique['track_artist_norm'].str.replace(r"feat\.?.*", "", regex=True)
spotify_en_unique['track_artist_norm'] = spotify_en_unique['track_artist_norm'].str.replace(r"featuring.*", "", regex=True)
spotify_en_unique['track_artist_norm'] = spotify_en_unique['track_artist_norm'].str.replace(r"ft\.?.*", "", regex=True)
spotify_en_unique['track_artist_norm'] = spotify_en_unique['track_artist_norm'].str.replace("&", "and", regex=False)
spotify_en_unique['track_artist_norm'] = spotify_en_unique['track_artist_norm'].str.replace(r"[^a-z0-9\s]", "", regex=True)
spotify_en_unique['track_artist_norm'] = spotify_en_unique['track_artist_norm'].str.replace(r"\s+", " ", regex=True).str.strip()

In [62]:
billboard_song_level['song_norm'] = billboard_song_level['song'].str.lower()
billboard_song_level['artist_norm'] = billboard_song_level['artist'].str.lower()

In [63]:
billboard_song_level['song_norm'] = billboard_song_level['song_norm'].str.replace(r"\(.*?\)", "", regex=True)
billboard_song_level['song_norm'] = billboard_song_level['song_norm'].str.replace(r"\[.*?\]", "", regex=True)
billboard_song_level['song_norm'] = billboard_song_level['song_norm'].str.replace(r"feat\.?.*", "", regex=True)
billboard_song_level['song_norm'] = billboard_song_level['song_norm'].str.replace(r"featuring.*", "", regex=True)
billboard_song_level['song_norm'] = billboard_song_level['song_norm'].str.replace(r"ft\.?.*", "", regex=True)
billboard_song_level['song_norm'] = billboard_song_level['song_norm'].str.replace(r"[^a-z0-9\s]", "", regex=True)
billboard_song_level['song_norm'] = billboard_song_level['song_norm'].str.replace(r"\s+", " ", regex=True).str.strip()

In [64]:
billboard_song_level['artist_norm'] = billboard_song_level['artist_norm'].str.replace(r"\(.*?\)", "", regex=True)
billboard_song_level['artist_norm'] = billboard_song_level['artist_norm'].str.replace(r"\[.*?\]", "", regex=True)
billboard_song_level['artist_norm'] = billboard_song_level['artist_norm'].str.replace(r"feat\.?.*", "", regex=True)
billboard_song_level['artist_norm'] = billboard_song_level['artist_norm'].str.replace(r"featuring.*", "", regex=True)
billboard_song_level['artist_norm'] = billboard_song_level['artist_norm'].str.replace(r"ft\.?.*", "", regex=True)
billboard_song_level['artist_norm'] = billboard_song_level['artist_norm'].str.replace("&", "and", regex=False)
billboard_song_level['artist_norm'] = billboard_song_level['artist_norm'].str.replace(r"[^a-z0-9\s]", "", regex=True)
billboard_song_level['artist_norm'] = billboard_song_level['artist_norm'].str.replace(r"\s+", " ", regex=True).str.strip()

In [65]:
spotify_billboard_match = spotify_en_unique.merge(
    billboard_song_level,
    left_on=['track_name_norm', 'track_artist_norm'],
    right_on=['song_norm', 'artist_norm'],
    how='left'
)

In [66]:
spotify_billboard_match['song'].notna().sum()

np.int64(3063)

In [67]:
spotify_billboard_match['is_charting'] = spotify_billboard_match['song'].notna().astype(int)
spotify_billboard_match['is_charting'].value_counts()

is_charting
0    10671
1     3063
Name: count, dtype: int64

In [68]:
matched = spotify_billboard_match['is_charting'].sum()
total = len(spotify_billboard_match)
print("Matched songs:", matched)
print("Total songs:", total)
print("Match percentage:", round(matched / total * 100, 2), "%")

Matched songs: 3063
Total songs: 13734
Match percentage: 22.3 %


In [69]:
spotify_billboard_match[spotify_billboard_match['is_charting'] == 1][
    ['track_name', 'track_artist', 'song', 'artist']
].head(20)

,track_name,track_artist,song,artist
1,Poison,Bell Biv DeVoe,Poison,Bell Biv DeVoe
6,Tender Lover,Babyface,Tender Lover,Babyface
9,Changes,2Pac,Changes,2Pac
11,Suga Suga,Baby Bash,Suga Suga,Baby Bash Featuring Frankie J
13,Secrets,The Weeknd,Secrets,The Weeknd
16,Get The Funk Out Ma Face,The Brothers Johnson,Get The Funk Out Ma Face,The Brothers Johnson
18,Juke Box Hero,Foreigner,Juke Box Hero,Foreigner
30,Phenomenal,Eminem,Phenomenal,Eminem
32,Try,P!nk,Try,P!nk
37,Love Sosa,Chief Keef,Love Sosa,Chief Keef


In [70]:
spotify_billboard_match[spotify_billboard_match['is_charting'] == 0][
    ['track_name', 'track_artist']
].head(20)

,track_name,track_artist
0,I Feel Alive,Steady Rollin
2,Baby It's Cold Outside (feat. Christina Aguilera),CeeLo Green
3,Dumb Litty,KARD
4,Soldier,James TW
5,Satisfy You,Diddy
7,Hide Away (feat. Envy Monroe),Blasterjaxx
8,Limestone,Magic City Hippies
10,Let Me Entertain You - Remastered 2011,Queen
12,You Don't Know Me - Radio Edit,Jax Jones
14,"Stand Out - From ""How to Build a Better Boy""",Sabrina Carpenter


After restricting the Spotify dataset to English-language songs with available lyrics and removing repeated playlist-based entries, 13,734 unique song-artist observations remained. A first exact-match merge with Billboard chart data, using normalized song and artist names, identified 3,063 songs that appeared in Billboard at least once. This corresponds to approximately 22.3% of the cleaned Spotify sample. The result suggests that the merged dataset is suitable for comparing charting and non-charting songs.